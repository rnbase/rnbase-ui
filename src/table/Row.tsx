import React from 'react';
import {
  StyleProp,
  StyleSheet,
  Text,
  TextStyle,
  TouchableHighlight,
  View,
  ViewProps,
} from 'react-native';

import { Theme, Themed, withTheme } from '../theming';

interface RowProps extends ViewProps {
  children?: React.ReactNode;
  title?: string;
  subtitle?: string;
  activeOpacity?: number;
  underlayColor?: string;
  titleStyle?: StyleProp<TextStyle>;
  subtitleStyle?: StyleProp<TextStyle>;
  onPress?: () => void;
}

const Row: React.FC<Themed<typeof createStyleSheet, RowProps>> = ({
  theme: { styles, colors },
  children,
  title,
  subtitle,
  activeOpacity = 1,
  underlayColor = colors.underlay,
  style,
  titleStyle,
  subtitleStyle,
  onPress,
  ...props
}) => {
  const titleText = title && (
    <Text key="title" style={[styles.title, titleStyle]} numberOfLines={1}>
      {title}
    </Text>
  );
  const subtitleText = subtitle && (
    <Text key="subtitle" style={[styles.subtitle, subtitleStyle]} numberOfLines={1}>
      {subtitle}
    </Text>
  );

  const label = !subtitle ? (
    titleText
  ) : (
    <View key="label" style={styles.label}>
      {titleText}
      {subtitleText}
    </View>
  );

  const content = (
    <View {...props} style={[styles.root, style]}>
      {label}
      {children}
    </View>
  );

  if (onPress) {
    return (
      <TouchableHighlight
        onPress={onPress}
        activeOpacity={activeOpacity}
        underlayColor={underlayColor}
      >
        {content}
      </TouchableHighlight>
    );
  }

  return content;
};

const createStyleSheet = ({ colors, fonts }: Theme) =>
  StyleSheet.create({
    root: {
      minHeight: 43,
      flexWrap: 'nowrap',
      flexDirection: 'row',
      alignItems: 'center',
      paddingVertical: 4,
      paddingHorizontal: 20,
      justifyContent: 'space-between',
    },
    label: {
      minHeight: 50,
      justifyContent: 'center',
    },
    title: {
      ...fonts.normal,
      fontSize: 17,
      color: colors.black,
    },
    subtitle: {
      ...fonts.normal,
      fontSize: 12,
      marginTop: 3,
      color: colors.gray,
    },
  });

export default withTheme(Row, createStyleSheet, 'Form.Row');
