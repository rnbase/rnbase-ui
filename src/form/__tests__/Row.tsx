import React from 'react';

// Note: test renderer must be required after react-native.
import TestRenderer from 'react-test-renderer';

const TouchableHighlight = (props: any) => React.createElement('TouchableHighlight', props);

jest.doMock(
  'react-native/Libraries/Components/Touchable/TouchableHighlight',
  () => TouchableHighlight,
);

/**
 * Under test
 */
import Row from '../Row';

type Props = React.ComponentProps<typeof Row>;

const Label = () => React.createElement('Label');
const Action = () => React.createElement('Action');
/**
 * Under test
 */
const createElement = (props: Partial<Props>) => (
  <Row {...props}>
    <Label />
    <Action />
  </Row>
);

const createRenderer = (props: Partial<Props>) => TestRenderer.create(createElement(props));

it('should render normally', () => {
  expect(createRenderer({})).toMatchSnapshot();
});

it('should render with custom styles', () => {
  expect(createRenderer({ style: { backgroundColor: 'red' } })).toMatchSnapshot();
});

it('should render as touchable', () => {
  expect(createRenderer({ onPress: jest.fn() })).toMatchSnapshot();
});

it('should handle onPress', () => {
  const onPress = jest.fn();
  const tree = createRenderer({ onPress });

  tree.root.findByType(TouchableHighlight).props.onPress();

  expect(onPress).toBeCalledWith();
});
