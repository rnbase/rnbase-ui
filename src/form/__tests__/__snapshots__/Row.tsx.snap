// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render as touchable 1`] = `
<TouchableHighlight
  activeOpacity={1}
  onPress={[MockFunction]}
  underlayColor="#0001"
>
  <View
    style={
      Array [
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "flexWrap": "nowrap",
          "justifyContent": "space-between",
          "minHeight": 48,
          "paddingHorizontal": 15,
        },
        undefined,
      ]
    }
  >
    <Label />
    <Action />
  </View>
</TouchableHighlight>
`;

exports[`should render normally 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "space-between",
        "minHeight": 48,
        "paddingHorizontal": 15,
      },
      undefined,
    ]
  }
>
  <Label />
  <Action />
</View>
`;

exports[`should render with custom styles 1`] = `
<View
  style={
    Array [
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "flexWrap": "nowrap",
        "justifyContent": "space-between",
        "minHeight": 48,
        "paddingHorizontal": 15,
      },
      Object {
        "backgroundColor": "red",
      },
    ]
  }
>
  <Label />
  <Action />
</View>
`;
